lib/fluent/plugin/buffer.rb

```diff
@@ -271,6 +312,18 @@ def enqueue_chunk(metadata)
         nil
       end
 
+      def enqueue_unstaged_chunk(chunk)
+        synchronize do
```

Comment:
All calls to operate `@queue` and `@stage` must be safe for any situation, including methods added in future. Otherwise, someday someone makes deadlock.